dispatch/dispatch.go

```diff
@@ -133,6 +138,134 @@ func (d *Dispatcher) run(it provider.AlertIterator) {
 	}
 }
 
+// AlertGroup represents how alerts alerts exist within an aggrGroup
+type AlertGroup struct {
+	Alerts   []*EnrichedAlert `json:"alerts"`
+	Labels   model.LabelSet   `json:"labels,omitempty"`
+	Receiver string           `json:"receiver,omitempty"`
+}
+
+// EnrichedAlert is the representation of an alert, with status info and
+// fingerprint.
+type EnrichedAlert struct {
+	*types.Alert
+	Status      types.AlertStatus
+	Fingerprint string
+	Receivers   *[]string
+}
+
+type AlertGroups []*AlertGroup
+
+func (ag AlertGroups) Swap(i, j int)      { ag[i], ag[j] = ag[j], ag[i] }
+func (ag AlertGroups) Less(i, j int) bool { return ag[i].Labels.Before(ag[j].Labels) }
+func (ag AlertGroups) Len() int           { return len(ag) }
+
+// Groups returns a slice of AlertGroups from the dispatcher's internal state.
+func (d *Dispatcher) Groups(matchers []*labels.Matcher, receiverFilter *regexp.Regexp, silenced, inhibited, active bool) AlertGroups {
+	groups := AlertGroups{}
+
+	d.mtx.RLock()
+	defer d.mtx.RUnlock()
+
+	seen := map[model.Fingerprint]*AlertGroup{}
+
+	// Keep a list of receivers for an alert to prevent checking each alert
+	// again against all routes. The alert has already matched against this
+	// route on ingestion.
+	receivers := map[model.Fingerprint][]string{}
+
+	for route, ags := range d.aggrGroups {
+		receiver := route.RouteOpts.Receiver
+		if receiverFilter != nil && !receiverFilter.MatchString(receiver) {
+			continue
+		}
+
+		for _, ag := range ags {
+			alertGroup, ok := seen[ag.fingerprint()]
+			if !ok {
+				alertGroup = &AlertGroup{
+					Labels:   ag.labels,
+					Receiver: receiver,
+				}
+
+				seen[ag.fingerprint()] = alertGroup
+			}
+
+			now := time.Now()
+
+			alerts := ag.alerts.List()
+			enrichedAlerts := make([]*EnrichedAlert, 0, len(alerts))
+			for a := range alerts {
+				if !a.EndsAt.IsZero() && a.EndsAt.Before(now) {
+					continue
+				}
+				fp := a.Fingerprint()
+				d.setAlertStatus(a.Labels)
+				status := d.marker.Status(fp)
+
+				if !active && status.State == types.AlertStateActive {
+					continue
+				}
+
+				if !silenced && len(status.SilencedBy) != 0 {
+					continue
+				}
+
+				if !inhibited && len(status.InhibitedBy) != 0 {
+					continue
+				}
+
+				if !matchesFilterLabels(a.Labels, matchers) {
+					continue
+				}
+
+				ea := &EnrichedAlert{
+					Alert:       a,
+					Status:      status,
+					Fingerprint: fp.String(),
+				}
+
+				if r, ok := receivers[fp]; ok {
+					// Receivers slice already exists. Add
+					// the current receiver.
+					// ea stores a pointer to this slice,
+					// so any changes to this slice will be
+					// available from ea.
+					r = append(r, receiver)
+				} else {
+					// First time we've seen this alert.
+					// Initialize a new receivers slice.
+					r := []string{receiver}
+					receivers[fp] = r
+					ea.Receivers = &r
```

Comment:
To get rid of the pointer to slice trick, this function could also return the `map[model.Fingerprint][]string` variable and let the caller look up the receiver list using the alert's fingerprint.